### PR TITLE
Files Quest

### PR DESCRIPTION
**Pull request by nginnever**
<h3>TODO</h3>
- [x] add CLI (tuples suplied to new awesomeness importer)
 - [ ] progress bar
 - [x] fix path names
- [x] add HTTP-API
- [x] cat CLI
 - [x] throw daemon not supported err
- [x] cat HTTP-API
- [x] get CLI
 - [ ] add progress bar
- [x] get HTTP-API
 - [ ] get JS-IPFS-API 
- [x] Fix add to respond to no path supplied
- [x] fix INIT to place config with new importer
- [x] Tests
 - [x] files add core
 - [x] files add cli
 - [x] files add http-api
 - [x] files cat core
 - [x] files cat cli
 - [x] files cat http-api
 - [x] files get core
 - [x] files get cli
 - [x] files get http-api

<h3>Blockers</h3>
- [x] unixfs-engine readable stream error <a href="https://github.com/ipfs/js-ipfs-unixfs-engine/issues/31">issue</a>

<h3>Call to Action</h3>
@bgrieder if you are still interested in the files commands that you were working we would love some help on this PR. Finally organized everything into a place that we can collab. This should close <a href="https://github.com/ipfs/js-ipfs/pull/104">pull 104</a>

@xicombd as I get into writing the http things I could use some help if possible. I hear you are pro at these things :D 

Let the quest begin!

### CODE REVIEW

File: src/cli/commands/files/cat.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    utils.getIPFS((err, ipfs) => {
+      if (err) {
+        throw err
+      }
+      if (utils.isDaemonOn()) {
+        ipfs.cat(path, (err, res) => {
+          if (err) {
+            throw new Error(err)
+          }
+          res.on('data', (data) => {
```
Comment (diasdavid) [inline src/cli/commands/files/cat.js]: you can just `res.pipe(process.stdout)`

File: test/core-tests/test-files.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+const streamifier = require('streamifier')
+const bs58 = require('bs58')
+
+const IPFS = require('../../src/core')
+
+describe('files', () => {
+  var ipfs
+
+  before((done) => {
+    ipfs = new IPFS()
```
Comment (diasdavid) [inline test/core-tests/test-files.js]: https://github.com/ipfs/js-ipfs/blob/master/test/core-tests/test-block.js#L22

**Comment by diasdavid** [comment]
@nginnever please update all the deps, unixfs-engine is 100% green now (node.js + browser). 

PS: would you mind squashing the commits made thus far :)?

File: src/cli/commands/files/add.js [hunk 3]
```diff
@@ -19,15 +24,120 @@ module.exports = Command.extend({
...
+    let rs
+
+    if (!path) {
+      throw new Error('Error: Argument \'path\' is required')
+    }
+
+    s = fs.statSync(path)
+
+    if (s.isDirectory() && recursive === false) {
+      throw new Error('Error: ' + process.cwd() + ' is a directory, use the \'-r\' flag to specify directories')
```
Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: Can you use backticks for these error messages, like this: 

```js
throw new Error(`Error: ${process.cwd()} is a directory, use the '-r' flag to specify directories`)
```

Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: Also I think the `Error:` in the front is not needed, in go-ipfs that's just the log level

File: src/cli/commands/files/add.js [hunk 4]
```diff
@@ -19,15 +24,120 @@ module.exports = Command.extend({
...
+      s = fs.statSync(process.cwd())
+      if (s.isDirectory()) {
+        throw new Error('Error: ' + process.cwd() + ' is a directory, use the \'-r\' flag to specify directories')
+      }
+    }
+
+    glob(pathj.join(path, '/**/*'), (err, res) => {
       if (err) {
-        return console.log(err)
+        throw new Error(err)
```
Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: Why does the `err` need to be wrapped? that should already be an instance of `Error`

File: src/cli/commands/files/add.js [hunk 5]
```diff
@@ -19,15 +24,120 @@ module.exports = Command.extend({
...
       }
-      console.log('added', bs58.encode(stats.Hash).toString(), stats.Name)
+      utils.getIPFS((err, ipfs) => {
+        if (err) {
+          throw new Error(err)
```
Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: Why does the `err` need to be wrapped? that should already be an instance of `Error`

File: src/cli/commands/files/add.js [hunk 6]
```diff
@@ -19,15 +24,120 @@ module.exports = Command.extend({
...
+                const buffered = fs.readFileSync(element)
+                rs.push(buffered)
+                rs.push(null)
+                const filePair = {path: addPath, content: rs}
+                files.push(filePair)
+                callback()
+              }
+            }, (err) => {
+              if (err) {
+                throw new Error(err)
```
Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: No need for wrapping the `err`

File: src/cli/commands/files/add.js [hunk 7]
```diff
@@ -19,15 +24,120 @@ module.exports = Command.extend({
...
+              }
+              ipfs.add(files, (err, res) => {
+                if (err) {
+                  throw new Error(err)
```
Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: No need for wrapping the `err`

File: src/cli/commands/files/add.js [hunk 8]
```diff
@@ -19,15 +24,120 @@ module.exports = Command.extend({
...
+            rs = new Readable()
+            const buffered = fs.readFileSync(path)
+            rs.push(buffered)
+            rs.push(null)
+            path = path.substring(path.lastIndexOf('/') + 1, path.length)
+            const filePair = {path: path, content: rs}
+            files.push(filePair)
+            ipfs.add(files, (err, res) => {
+              if (err) {
+                throw new Error(err)
```
Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: No need for wrapping the `err`

File: src/cli/commands/files/add.js [hunk 9]
```diff
@@ -19,15 +24,120 @@ module.exports = Command.extend({
...
+              const filePair = {path: addPath, stream: rs}
+              i.write(filePair)
+              callback()
+            }
+          }, (err) => {
+            if (err) {
+              throw new Error(err)
+            }
+            i.end()
+            return
```
Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: no need for `return`

File: src/cli/commands/files/cat.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+      options = {}
+    }
+    utils.getIPFS((err, ipfs) => {
+      if (err) {
+        throw err
+      }
+      if (utils.isDaemonOn()) {
+        ipfs.cat(path, (err, res) => {
+          if (err) {
+            throw new Error(err)
```
Comment (dignifiedquire) [inline src/cli/commands/files/cat.js]: No need for wrapping the `err`

File: src/cli/commands/files/cat.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+          }
+          console.log(res.toString())
+        })
+        return
+      }
+
+      ipfs.files.cat(path, (err, res) => {
+        if (err) {
+          throw new Error(err)
```
Comment (dignifiedquire) [inline src/cli/commands/files/cat.js]: No need for wrapping the `err`

File: src/cli/commands/files/get.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+      dir = options
+    }
+
+    utils.getIPFS((err, ipfs) => {
+      if (err) {
+        throw err
+      }
+      ipfs.files.get(path, (err, data) => {
+        if (err) {
+          throw new Error(err)
```
Comment (dignifiedquire) [inline src/cli/commands/files/get.js]: No need for wrapping the `err`

File: src/cli/commands/files/get.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+          } else {
+            filepath = data.path.substring(0, data.path.lastIndexOf('/') + 1)
+            try {
+              fs.mkdirSync(dir + filepath)
+            } catch (err) {
+            }
+            ws = fs.createWriteStream(dir + data.path)
+            // data.stream.on('end', () => {
+            //  console.log('finished writing file to disk')
+            // })
```
Comment (dignifiedquire) [inline src/cli/commands/files/get.js]: why is this commented out?

File: src/cli/commands/files/get.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+              try {
+                fs.mkdirSync(dir + filepath)
+              } catch (err) {
+                console.log(err)
+              }
+            }
+          } else {
+            filepath = data.path.substring(0, data.path.lastIndexOf('/') + 1)
+            try {
+              fs.mkdirSync(dir + filepath)
```
Comment (dignifiedquire) [inline src/cli/commands/files/get.js]: maybe `path.join` is safer?

File: src/cli/commands/files/get.js [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+            } catch (err) {
+            }
+            ws = fs.createWriteStream(dir + data.path)
```
Comment (dignifiedquire) [inline src/cli/commands/files/get.js]: maybe `path.join` is safer?

File: src/cli/commands/files/get.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+              } catch (err) {
+                console.log(err)
+              }
+            }
+          } else {
+            filepath = data.path.substring(0, data.path.lastIndexOf('/') + 1)
+            try {
+              fs.mkdirSync(dir + filepath)
+            } catch (err) {
+            }
```
Comment (dignifiedquire) [inline src/cli/commands/files/get.js]: why can this `err` be ignored?

File: src/core/index.js [hunk 17]
```diff
@@ -392,10 +394,52 @@ function IPFS (repo) {
...
+      }
+
+      const i = new Importer(dagS)
+      const res = []
+
+      i.on('data', (info) => {
+        res.push(info)
+      })
+
+      i.on('end', () => {
```
Comment (dignifiedquire) [inline src/core/index.js]: Use `once` it's safer

File: src/core/init.js [hunk 18]
```diff
@@ -63,17 +67,48 @@ module.exports = (repo, opts, callback) => {
...
     const dag = new DagService(blocks)
 
     const initDocsPath = path.join(__dirname, '../init-files/init-docs')
 
-    importer.import(initDocsPath, dag, {
-      recursive: true
-    }, doneImport)
+    const i = new Importer(dag)
+    i.on('data', (file) => {
+    })
```
Comment (dignifiedquire) [inline src/core/init.js]: why an empty listener?

File: src/core/init.js [hunk 19]
```diff
@@ -63,17 +67,48 @@ module.exports = (repo, opts, callback) => {
...
+          const filePair = {path: addPath, stream: rs}
+          i.write(filePair)
+          callback()
+        }
+      }, (err) => {
+        if (err) {
+          throw err
+        }
+        i.end()
+        return
```
Comment (dignifiedquire) [inline src/core/init.js]: no need for `return`

**Comment by diasdavid** [comment]
@voxelot can you break this PR into two that

- 1st - Implement Files at the IPFS Core Level and CLI Offline 
- 2nd - Implement Files with js-ipfs daemon on, ensuring its compatibility with go-ipfs, this means:
  - go-ipfs cli -> js-ipfs http api
  - js-ipfs cli -> go-ipfs http api

I'm requesting this because it seems that you are close, or even 100% done for the 1st and still a bit far from having the 2nd 100%. Currently, not having the 1st complete is blocking the pursuit of js-ipfs running on the browser with the Files API.

**Comment by nginnever** [comment]
Closing this in favor of new HTTP branch and completion of CLI PR <a href="https://github.com/ipfs/js-ipfs/pull/204">204</a>

**Pull request closed without merging**